graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -135,6 +139,7 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
         this.rulesList.addAll(smallerRules);
         this.rulesList.addAll(containsRules);
         this.rulesList.addAll(regexRules);
+        this.rulesList.addAll(matchInputRules);
```

Comment:
The rules list should be ordered by matching complexity of the rules (ascending), so the match input rules should be before or after `exactRules`.

Reply:
I moved it below `contains` rule. Since it is just that.

Reply:
It should be after the `exactRules`, as it does not do a `contains` check, but an `equals` check.